torch/distributed/rpc/backend_registry.py

```diff
@@ -260,55 +334,118 @@ def _tensorpipe_init_backend_handler(store, name, rank, world_size, rpc_backend_
+            # process initializes its PyTorch CUDA states.
+            torch.cuda.init()
+            device_count = torch.cuda.device_count()
+        else:
+            device_count = 0
+
+        reverse_device_maps, devices = _tensorpipe_exchange_and_check_all_device_maps(
+            name,
+            device_count,
+            rpc_backend_options.device_maps,
+            rpc_backend_options.devices,
+            group,
+        )
 
-    api._init_rpc_states(agent)
+        # TODO: add try-except and destroy _agent in all processes if any fails.
+        agent = TensorPipeAgent(
+            store,
+            name,
+            rank,
+            world_size,
+            rpc_backend_options,
+            reverse_device_maps,
+            devices,
+        )
 
-    # Run one dummy round of RPC to initialize channels/transports. Without
-    # this, it's easy to hit timeout in rpc.shutdown() if there is no other RPC
-    # on that process before rpc.shutdown(), as the agent initialization can
-    # take longer than 5s.
-    api._all_gather(None, timeout=rpc_backend_options.rpc_timeout)
-    # Need a barrier here to make sure no peers leave before the rank0 finishes
-    # _all_gather
-    group.barrier().wait()
+        api._init_rpc_states(agent)
 
-    return agent
+        # Run one dummy round of RPC to initialize channels/transports. Without
+        # this, it's easy to hit timeout in rpc.shutdown() if there is no other RPC
+        # on that process before rpc.shutdown(), as the agent initialization can
+        # take longer than 5s.
+        api._all_gather(None, timeout=rpc_backend_options.rpc_timeout)
+        # Need a barrier here to make sure no peers leave before the rank0 finishes
+        # _all_gather
+        group.barrier().wait()
 
+        return agent
+    # initialization for dynamic rpc (ranks can join and leave)
+    else:
+        # TODO: retrieve token from store to signal start of rank join/leave critical section
+        token = f"TokenOnWorker{rank}"
+
+        while True:
+            returned = store.compare_set("init_rpc_token", "", token).decode()
+            if returned == token:
+                print(f"{rank} got token")
+                if torch.cuda.is_available():
+                    # It's necessary to initialize PyTorch CUDA states here (e.g.,
+                    # CUDACachingAllocator). If this is missing, we could hit errors like
+                    # "allocator not initialized", because other processes might send
+                    # CUDA-related RPC request to this process before user code in this
+                    # process initializes its PyTorch CUDA states.
+                    torch.cuda.init()
+                    device_count = torch.cuda.device_count()
+                else:
+                    device_count = 0
+
+                # Validate devices and device_maps locally for current rank
+                _tensorpipe_check_local_device_maps(name, rpc_backend_options)
```

Comment:
we can do all of these before grabbing the token to minimize the contention for the critical section?